src/app/interfaces/vm-device.interface.ts

```diff
@@ -3,104 +3,40 @@ import {
   VmDeviceType, VmDiskMode, VmDisplayType, VmNicType,
 } from 'app/enums/vm.enum';
 
-interface VmPciPassthroughAttributes {
-  pptdev: string;
-  type: string;
-}
-
-interface VmUsbPassthroughAttributes {
-  controller_type: string;
-  device: string | null;
-  usb?: {
-    product_id?: string;
-    vendor_id?: string;
-  };
-}
-
-export interface VmDisplayAttributes {
-  bind: string;
-  password: string;
-  password_configured?: boolean;
-  port: number;
-  resolution: string;
-  type: VmDisplayType;
-  wait: boolean;
-  web: boolean;
-}
-
-export interface VmCdRomAttributes {
-  path: string;
-}
-
-export interface VmRawFileAttributes {
-  boot: boolean;
-  logical_sectorsize: number;
-  path: string;
-  physical_sectorsize: number;
-  size: number;
-  type: VmDiskMode;
-}
-
-export interface VmNicAttributes {
-  mac: string;
-  nic_attach: string;
-  type: VmNicType;
-  trust_guest_rx_filters: boolean;
-}
-
-export interface VmDiskAttributes {
-  logical_sectorsize: number;
-  path: string;
-  physical_sectorsize: number;
-  type: VmDiskMode;
-
-  // TODO: May be only relevant when creating a vm
-  create_zvol?: boolean;
-  zvol_name?: string;
-  zvol_volsize?: number;
-}
-
-export interface BaseVmDevice {
+export interface BaseVmDevice<T> {
   id: number;
   dtype: VmDeviceType;
-  attributes: unknown;
+  attributes: T;
   order: number;
   vm: number;
 }
 
-export interface VmPciPassthroughDevice extends BaseVmDevice {
+export interface VmPciPassthroughDevice extends BaseVmDevice<VmPciPassthroughAttributes> {
   dtype: VmDeviceType.Pci;
-  attributes: VmPciPassthroughAttributes;
-}
-
-export interface VmRawFileDevice extends BaseVmDevice {
-  dtype: VmDeviceType.Raw;
-  attributes: VmRawFileAttributes;
 }
 
-export interface VmNicDevice extends BaseVmDevice {
-  dtype: VmDeviceType.Nic;
-  attributes: VmNicAttributes;
+export interface VmUsbPassthroughDevice extends BaseVmDevice<VmUsbPassthroughAttributes> {
+  dtype: VmDeviceType.Usb;
 }
 
-export interface VmDisplayDevice extends BaseVmDevice {
+export interface VmDisplayDevice extends BaseVmDevice<VmDisplayAttributes> {
   dtype: VmDeviceType.Display;
```

Comment:
You have `dtype` defined in two places now, here and in attributes.
Also instead of passing attributes through generic, you should be able to do:
```
export interface VmDisplayDevice extends BaseVmDevice {
   attributes: VmDisplayAttribute;
}
```

Reply:
sorry, missed that `dtype`, will remove it.
Yeah instead of generics for sure me way use syntax you provided 👍